subsys: logging: Add MIPI sys-t support for v2 logging subsystem.

aasthagr opened the pull request:
Adding functions log_output_msg2_syst_process and hexdump2_print to support v2 logging subsystem. Also adding v2 logging support for sys-t sample.

aasthagr committed:
logging: Add MIPI sys-t support for v2 logging subsystem.

Adding functions log_output_msg2_syst_process and hexdump2_print
to support v2 logging subsystem.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

peter-mitsis commented on subsys/logging/log_output_syst.c:
```diff
@@ -672,17 +673,84 @@ void log_output_string_syst_process(const struct log_output *log_output,
-				struct log_msg_ids src_level,
-				const uint8_t *data, uint32_t length, uint32_t flag)
+	buf = log_msg2_get_data(msg, &length);
+	MIPI_SYST_WRITE(&log_syst_handle, severity, 0x1A, buf, length);
```
It is unusual to see a hard coded value that is neither 0 nor 1.
1. What is the purpose of 0x1A?
2. If it is correct, can we use a descriptive macro?

aasthagr replied:
@peter-mitsis  just following the semantics of MIPI_SYS_WRITE used for v1 logging implementation for now.

ceolin commented on subsys/logging/log_backend_net.c:
```diff
@@ -199,6 +199,8 @@ static void process(const struct log_backend *const backend,
 	uint32_t flags = LOG_OUTPUT_FLAG_FORMAT_SYSLOG | LOG_OUTPUT_FLAG_TIMESTAMP;
 
+	flags |= IS_ENABLED(CONFIG_LOG_BACKEND_UART_SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
+
```
are you sure about it ? This is confuse, because you basically enabling two different formats in this bitmask,
`FORMAT_SYSLOG` and `FORMAT_SYST`.

aasthagr replied:
Have removed this from log_backend_net.c because got to know that SYS-T is not supported for network backend.

ceolin commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -320,6 +320,8 @@ static void process(const struct log_backend *const backend,
 {
 	uint32_t flags = log_backend_std_get_flags();
 
+	flags |= IS_ENABLED(CONFIG_LOG_BACKEND_UART_SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
```
this is not the right place for this. You should move this check and modify the flags in `log_backend_std_get_flags()`

aasthagr replied:
@ceolin  std_get_flags() routine is common for all backends and incase 2 backends are enabled at same time, we want to make sure setting of syst format happens outside the std_get_flags() function.

ceolin commented on subsys/logging/log_backend_swo.c:
```diff
@@ -87,6 +87,8 @@ static void log_backend_swo_process(const struct log_backend *const backend,
 	uint32_t flags = log_backend_std_get_flags();
 
+	flags |= IS_ENABLED(CONFIG_LOG_BACKEND_UART_SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
+
```
ditto

ceolin commented on subsys/logging/log_backend_uart.c:
```diff
@@ -65,6 +65,7 @@ static void process(const struct log_backend *const backend,
 {
 	uint32_t flags = log_backend_std_get_flags();
 
+	flags |= IS_ENABLED(CONFIG_LOG_BACKEND_UART_SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
```
ditto

ceolin commented on subsys/logging/log_output_syst.c:
```diff
@@ -17,6 +17,8 @@
 static struct mipi_syst_header log_syst_header;
 static struct mipi_syst_handle log_syst_handle;
 
+#define DATA_ID 0x1A
```
Can you add some comment about what this value means ?

aasthagr replied:
This is the data id which is a numeric value in the range of 0..31 to identify the content pointed to by data. I have removed the macro for now to be consistent with how MIPI_SYST_WRITE was used in v1 logging implementation.

ceolin commented on subsys/logging/log_output_syst.c:
```diff
@@ -672,17 +675,84 @@ void log_output_string_syst_process(const struct log_output *log_output,
+	MIPI_SYST_WRITE(&log_syst_handle, severity, DATA_ID, buf, length);
+}
 
-void log_output_hexdump_syst_process(const struct log_output *log_output,
```
You can't remove it. It is used by LOG V1 syst implementation. `log_output_hexdump` calls it.

aasthagr replied:
It was by accident I think. Have put this back.

ceolin reviewed (changes requested):
There are things to fix, please ensure that you don't break V1 implementation.

dcpleung commented:
For the sample, you need to *add* v2 support and *not replacing* v1.

aasthagr commented:
> There are things to fix, please ensure that you don't break V1 implementation.

@ceolin made some changes and addressed the comments, can you please re-review? Thanks!!

dcpleung commented on samples/subsys/logging/syst/prj.conf:
```diff
@@ -18,3 +18,4 @@ CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=100
 CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD=2
 CONFIG_LOG_PRINTK_MAX_STRING_LENGTH=256
 CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=4096
+CONFIG_LOG2_MODE_DEFERRED=y
```
You are changing this test to do v2 logging, and v1 is no longer being tested.

aasthagr replied:
Yeahh you are right. Removed the the v2 logging config from prj.conf and added support for testing sys-t sample with v2 logging subsystem as well in sample.yaml. Thanks!!

nordic-krch commented on subsys/logging/log_output_syst.c:
```diff
@@ -684,6 +685,85 @@ void log_output_hexdump_syst_process(const struct log_output *log_output,
+		fmt = ((char **)buf)[1];
+		args_size = ((((uint8_t *)buf)[0] * sizeof(int))/4)-2;
+		buf += sizeof(char *) * 2;
+
```
Do you know if it would be possible to extend mipi syst library and add `MIPI_SYS_VPRINTF(h, sev, fmt, ap)`? It would be rather trivial change but i don't how library is supported.

That would help here a lot as after getting `ap` and `fmt` from cbprintf package you would directly call `MIPI_SYS_VPRINTF`.

Current solution has severe limitations, up to 4 args and only `int` args.

aasthagr replied:
@nordic-krch That is a very good point. MIPI sys-t doesn't support any VPRINTF type API AFAIK. But have added upto 15 switch cases to extend the support. It can only support int args for now (Note: v1 logging in general and with sys-t as well also doesn't support float values). I can add the support for other parameter types as a followup improvement to this. Let me know your thoughts.

nordic-krch replied:
what about adding vprintf support to mipi sys-t. I see that we are maintaining a zephyr fork so it would be straightforward to add it there. Don't know who maintains upstream project but since it is not very controversial change i guess that we could upstream it as well. That would solve all the issue. @nashif do you know if upstream project is opened for patches like that?

nashif replied:
will check

aasthagr replied:
@nordic-krch have the VPRINTF supported now [here](https://github.com/zephyrproject-rtos/zephyr/pull/41277). This will also support floating point variables alongwith printing variable number of arguments. Please re-review and let me know if any other changes are needed.

dcpleung commented on samples/subsys/logging/syst/sample.yaml:
```diff
@@ -1,11 +1,23 @@
       type: one_line
       regex:
         - "SYS-T RAW DATA: "
+    extra_configs:
+      - CONFIG_LOG2_MODE_DEFERRED=n
```
This should not need to be disabled. v2 logging is not enabled by default.

aasthagr replied:
Removed the disabling for v1 logging in sample.yaml.

aasthagr committed:
samples: logging: Add v2 logging support for sys-t sample.

Adding #ifndef CONFIG_LOG2 condition for log_msg_ids and
log_string_sync in sys-t sample. Extending support for
v2 logging in sys-t sample.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nordic-krch commented:
Is this one superseded by #41277?

aasthagr commented:
> Is this one superseded by #41277?

Yes #41277 contains changes from this PR along with the support for MIPI_SYST_VPRINTF. I can close this one.
